module/modmanager/manager.go

```diff
@@ -349,6 +362,150 @@ func (mgr *Manager) getModule(conf resource.Config) (*module, bool) {
 	return nil, false
 }
 
+var (
+	// oueTimeout is the length of time for which an OnUnexpectedExit function
+	// can execute blocking calls.
+	oueTimeout = 2 * time.Minute
+	// oueRestartInterval is the interval of time at which an OnUnexpectedExit
+	// function can attempt to restart the module process. Multiple restart
+	// attempts will use basic backoff.
+	oueRestartInterval = 5 * time.Second
+)
+
+// newOnUnexpectedExitHandler returns the appropriate OnUnexpectedExit function
+// for the passed-in module to include in the pexec.ProcessConfig.
+func (mgr *Manager) newOnUnexpectedExitHandler(mod *module) func(exitCode int) bool {
+	return func(exitCode int) bool {
+		mod.inRecoveryLock.Lock()
+		defer mod.inRecoveryLock.Unlock()
+		if mod.inRecovery.Load() {
+			return false
+		}
+		mod.inRecovery.Store(true)
+		defer mod.inRecovery.Store(false)
+
+		// Use oueTimeout for entire attempted module restart.
+		ctx, cancel := context.WithTimeout(context.Background(), oueTimeout)
+		defer cancel()
+
+		// Log error immediately, as this is unexpected behavior.
+		mgr.logger.Errorw(
+			"module has unexpectedly exited, attempting to restart it",
+			"module", mod.name,
+			"exit_code", exitCode,
+		)
+
+		// If attemptRestart returns any orphaned resource names, restart failed,
+		// and we should remove orphaned resources. Since we handle process
+		// restarting ourselves, return false here so goutils knows not to attempt
+		// a process restart.
+		if orphanedResourceNames := mgr.attemptRestart(ctx, mod); orphanedResourceNames != nil {
+			mgr.removeOrphanedResources(ctx, orphanedResourceNames)
```

Comment:
This is calling a variable field that could be nil. Even if guards are added on creation, should check for nil here as well.